exporter/__init__.py

```diff
@@ -0,0 +1,283 @@
+import time
+import os
+import sys
+import signal
+import logging
+
+import argparse
+import collections
+from typing import List, Dict
+
+from threading import Thread
+from queue import Queue
+
+from celery import Celery, Task as CeleryTask
+import celery.states
+import celery.events
+from celery.utils.objects import FallbackContext
+
+from kombu.exceptions import TimeoutError as QueueIsEmptyError
+from amqp.exceptions import ChannelError as QueueNotFoundError
+
+from prometheus_client import Histogram, Gauge, start_http_server
+from metrics import (
+    TasksByStateGauge,
+    TasksByStateAndNameGauge,
+    TaskRuntimeByNameHistogram,
+    WorkersGauge,
+    # TaskLatencyHistogram,
+    QueueLengthGauge,
+    registry,
+)
+
+import amqp.exceptions
+
+__VERSION__ = open("VERSION", "r").read().strip()
+
+LOG_FORMAT = "[%(asctime)s] %(name)s:%(levelname)s: %(message)s"
+
+PRIORITY_STEPS = [0, 3, 6, 9]
+PRIORITY_SEPARATOR = "\x06\x16"
+
+
+class CeleryEventReceiverThread(Thread):
+    def __init__(self, celery_app: Celery, task_state, event_queue: Queue):
+        Thread.__init__(self, daemon=True)
+        self.celery_app = celery_app
+        self.task_state = task_state
+        self.event_queue = event_queue
+
+    def run(self):
+        with self.celery_app.connection() as connection:
+            receiver = self.celery_app.events.Receiver(
+                connection,
+                handlers={
+                    "*": self.handle_event,
+                },
+            )
+            receiver.capture(limit=None, timeout=None, wakeup=True)
+
+    def handle_event(self, event: Dict[str, str]):
+        self.task_state.event(event)
+        self.collect_task_metrics(event)
+        # sleep so main thread can read new state
+        time.sleep(0.01)
```

Comment:
Should we make this configurable? Every 10ms potentially seems quite aggressive

Reply:
Ah never mind, I realise the metrics collection interval is configurable separately below

Reply:
Its a good point though we should really reduce this so it can do higher throughput.